scripts/build_images/build.py

```diff
@@ -0,0 +1,205 @@
+import hashlib
+import json
+import os
+import subprocess
+import tempfile
+from contextlib import nullcontext
+
+from .utils import get_env_or_exit, local_registry, print_and_run_command
+
+# The container name for the local Docker registry.
+REGISTRY_NAME = "prairielearn-registry"
+
+# A mapping from images to their base images. We use this to know to rebuild images when their base image has changed.
+BASE_IMAGE_MAPPING = {
+    # Core images.
+    "prairielearn/prairielearn": "prairielearn/plbase",
+    "prairielearn/executor": "prairielearn/prairielearn",
+    # Workspace images.
+    "prairielearn/workspace-vscode-python": "prairielearn/workspace-vscode-base",
+    "prairielearn/workspace-vscode-cpp": "prairielearn/workspace-vscode-base",
+}
+
+BASE_IMAGES = list(set(BASE_IMAGE_MAPPING.values()))
+
+images = get_env_or_exit("IMAGES")
+tag = get_env_or_exit("TAG")
+metadata_dir = os.environ.get("METADATA_DIR")
+should_push = os.environ.get("PUSH_IMAGES", "false").lower() == "true"
+
+if not metadata_dir:
+    print("No manifest directory specified; manifests will not be created.")
+
+
+def get_image_path(image: str) -> str:
+    if not image.startswith("prairielearn/"):
+        raise ValueError(f"Cannot build non-PrairieLearn image: {image}")
+
+    image = image[len("prairielearn/") :]
+
+    if image.startswith("workspace-"):
+        image = image[len("workspace-") :]
+        return f"workspaces/{image}"
+
+    if image.startswith("grader-"):
+        image = image[len("grader-") :]
+        return f"graders/{image}"
+
+    if image == "plbase":
+        return "images/plbase"
+
+    if image == "executor":
+        return "images/executor"
+
+    if image == "prairielearn":
+        return "."
+
+    raise ValueError(f"Cannot build unknown image: {image}")
+
+
+def get_current_platform() -> str:
+    result = subprocess.run(
+        ["docker", "version", "--format", "json"],
+        capture_output=True,
+        check=True,
+    )
+    version_data = json.loads(result.stdout)
+    return f"{version_data['Server']['Os']}/{version_data['Server']['Arch']}"
+
+
+def check_path_modified(path: str) -> bool:
+    branch = (
+        subprocess.run(
+            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
+            capture_output=True,
+            check=True,
+        )
+        .stdout.decode()
+        .strip()
+    )
+
+    # If this script is being run *on* the master branch, then we want to diff
+    # with the previous commit on master. Otherwise, we diff with master itself.
+    diff_branch = "HEAD^1" if branch == "master" else "remotes/origin/master"
```

Comment:
If I have stacked branches `master -> A -> B`, this diffs `B` with `master`. Is that intended?

Reply:
I didn't think about that case explicitly, but I don't think it's necessarily bad behavior. We can change the diff target in Actions if you want to handle this case.

Reply:
I think this is a niche enough case that it doesn't matter. But maybe a comment in the script describing that this happens.